IMN 262 - implement & test all file manager usages in catalog

ecamellini opened the pull request:
Closes [IMN-262](https://pagopa.atlassian.net/browse/IMN-262)

# Description

This PR:
* Fixes the Clone Descriptor implementation, where filemanager was not called correctly
* Adds tests for all calls involving filemanager, actually testing the filemanager integration
* **NOTE: there is another test that must be added when https://github.com/pagopa/interop-be-monorepo/pull/204 is merged, it will be done in a further PR** (tracked here https://pagopa.atlassian.net/browse/IMN-275)

# Tests

## 1) Automated tests

Covering file manager integration in all tests that involve file manager calls - new automated tests work ✅

## 2) Manual test - Clone Descriptor 

### **Descriptor & MinIO before cloning:**


<img width="934" alt="Screenshot 2024-02-20 at 17 19 13" src="https://github.com/pagopa/interop-be-monorepo/assets/6418684/081747bb-d29f-429c-bdb4-a498383a8842">
<img width="705" alt="Screenshot 2024-02-20 at 17 19 32" src="https://github.com/pagopa/interop-be-monorepo/assets/6418684/44cc2a08-50b4-4b46-998d-27bbb94f237d">
<img width="736" alt="Screenshot 2024-02-20 at 17 18 31" src="https://github.com/pagopa/interop-be-monorepo/assets/6418684/e0caf885-eca0-4d00-bad7-51bd1520d121">

### **Cloning succeeds** and we can see new interface and documents cloned

<img width="686" alt="Screenshot 2024-02-20 at 17 24 52" src="https://github.com/pagopa/interop-be-monorepo/assets/6418684/8bc8bade-17a2-4043-826f-058aa7ba9c59">

### **Descriptor & MinIO after cloning:** 

<img width="919" alt="Screenshot 2024-02-20 at 17 19 54" src="https://github.com/pagopa/interop-be-monorepo/assets/6418684/f4ae8ee0-d0bc-403f-9fd5-b70bed6f187d">
<img width="679" alt="Screenshot 2024-02-20 at 17 19 49" src="https://github.com/pagopa/interop-be-monorepo/assets/6418684/a3493131-da91-40e4-92b8-1d313b61dc48">
<img width="1700" alt="Screenshot 2024-02-20 at 17 20 03" src="https://github.com/pagopa/interop-be-monorepo/assets/6418684/b0fca981-9745-47cc-ae9a-d23fe4cbcdc9">

## 3) Manual test - Delete Descriptor

### Deleting the descriptor I just created by cloning
<img width="714" alt="Screenshot 2024-02-20 at 17 29 17" src="https://github.com/pagopa/interop-be-monorepo/assets/6418684/bbf145cf-0642-4d83-b2c6-1ed7f519d175">

### Descriptor & MinIO after cloning

<img width="957" alt="image" src="https://github.com/pagopa/interop-be-monorepo/assets/6418684/ae2e6c62-6841-46be-a5e2-a1597ae284e1">

<img width="1698" alt="Screenshot 2024-02-20 at 17 29 32" src="https://github.com/pagopa/interop-be-monorepo/assets/6418684/d8c21aa0-5054-4da5-ac9e-b574d44d4b04">



[IMN-262]: https://pagopa.atlassian.net/browse/IMN-262?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

ecamellini commented on packages/catalog-process/src/services/catalogService.ts:
```diff
@@ -1184,40 +1197,44 @@ export async function cloneDescriptorLogic({
   }
 
   const descriptor = retrieveDescriptor(descriptorId, eService);
-  const sourceDocument = descriptor.docs[0];
```
This was a bug: it was cloning the interface using the first document as data source, and not the interface

MalpenZibo reviewed (approved):
:ok_hand:

taglioni-r commented on packages/catalog-process/test/catalogService.integration.test.ts:
```diff
@@ -1547,16 +1702,75 @@ describe("database test", async () => {
           ...eService,
           id: unsafeBrandId(writtenPayload.eService!.id),
-          name: `${eService.name} - clone - 1/1/2024 12:00:00`,
+          name: `${eService.name} - clone - ${expectedLocalDate} ${expectedLocalTime}`,
```
Could we revert this change? It was specifically written in that way to have a clean overview of the expected formatting.

MalpenZibo replied:
Yes but in my local environment, it makes the test fail. I think that not in every environment `.toLocaleDateString("it-IT")` produces that result.

MalpenZibo replied:
It also causes a problem with `minio`. If we set the backend timezone then we get this error
```
- ApiError during EService creation with name eService nam
e - clone - 1/1/2024 12:00:00
+ The difference between the request time and the server's
 time is too large.
 ```

ecamellini replied:
Yes, unfortunately reverting the change breaks the integration tests with MinIO - and it's not straightforward to change the date of the MinIO container to set it to the same date we set here. 
If it is not strictly functional for the test, I would prefer keeping it dynamic like this. Otherwise, I can try to investigate a way to change the date of a single Docker container

taglioni-r replied:
Based on your points, for me it's not a problem to keep your changes. My concerns were based on [this](https://github.com/pagopa/interop-be-monorepo/pull/174#discussion_r1477983056). Let's see if @galales agrees.

Anyways, if I understand correctly, the problem we are raising now is that we have no clear assurance of how the name will be built. For example, in my local environment (and in the pipeline) it is `${eService.name} - clone - 1/1/2024 12:00:00`, whereas in @MalpenZibo's environment is something like `${eService.name} - clone - 01/01/2024 12:00:00` or similar? (In each environment now it passes because the logic builds the string just like it is expected from the tests, but there might be no consistency)

ecamellini replied:
Yes, that's one problem :)
The other problem is that MinIO mimics S3 and S3 requires the requester and the server to be in sync time-wise, otherwise it fails with: `S3 Error: The difference between the request time and the current time is too large`.
So, if we set the date for the tests with `vi.setSystemTime(new Date("2024-01-01T12:00:00"));`, we must somehow sync MinIO with that, which is not straightforward (it would mean setting the system time of the container in some way).

Maybe to improve readability of the string format in this test, we could add a comment above?
For example `//  EService Name - clone - 1/1/2024 12:00:00`

taglioni-r replied:
> Yes, that's one problem :) The other problem is that MinIO mimics S3 and S3 requires the requester and the server to be in sync time-wise, otherwise it fails with: `S3 Error: The difference between the request time and the current time is too large`.
So, if we set the date for the tests with `vi.setSystemTime(new Date("2024-01-01T12:00:00"));`, we must somehow sync MinIO with that, which is not straightforward (it would mean setting the system time of the container in some way).

I know, I got it. For what concerns this problem, for me it's ok to keep the dynamic date.

> Maybe to improve readability of the string format in this test, we could add a comment above? For example `// EService Name - clone - 1/1/2024 12:00:00`

This doesn't solve the inconsistency. The problem is that the expected string is based on the environment (`1/1/2024` or `01/01/2024`). Maybe we should find a way to build the string in the same way for every environment.

galales replied:
> Yes but in my local environment, it makes the test fail. I think that not in every environment .toLocaleDateString("it-IT") produces that result.

This is the reason why the test was explicit.
I do not have strong opinions about the format itself (even though I prefer `01/01/2024 12:00:00`), but it should be at least consistent. 
If we cannot achieve consistency with `.toLocaleDateString("it-IT")` then we probably need to build the date in a different way.

In tests we can re-build the date, being able to avoid mocking the time and having issues with MinIO.

One concern: is it me or this test smells flakiness? If timestamp taken from the test and timestamp taken from the logic falls in two different seconds, wouldn't this test fail?

taglioni-r replied:
> One concern: is it me or this test smells flakiness? If timestamp taken from the test and timestamp taken from the logic falls in two different seconds, wouldn't this test fail?

@galales what happens with the dynamic date creation is the following:
- on my local environment (and in the pipeline) the logic creates the date in `versionA` and the test expects it in `versionA`
- in @MalpenZibo's local environment the logic creates the date in `versionB` and the test expects it in `versionB`

`versionA`: `1/1/2024 12:00:00`
`versionB`: `01/01/2024 12:00:00`

galales replied:
yup, that's why I suggested this 
> If we cannot achieve consistency with .toLocaleDateString("it-IT") then we probably need to build the date in a different way.

meaning that we can avoid relying on that function, let's use something where we can explicitly define the format (which to me seems both clear and consistent)

ecamellini replied:
Roger, I will try to explicitly define the formatted date so that we do not rely on `toLocaleDateString("it-IT")`

ecamellini replied:
Roger, I will try to explicitly define the formatted date so that we do not rely on toLocaleDateString("it-IT").

For what concerns the test flakiness, I would use a regex at least on the seconds, so that we do not check for an exact second match. Thoughts?

taglioni-r commented on packages/catalog-process/test/catalogService.integration.test.ts:
```diff
@@ -1572,8 +1786,10 @@ describe("database test", async () => {
 
-        const conflictEServiceName = `${eService1.name} - clone - 1/1/2024 12:00:00`;
-
+        const cloneTimeStamp = new Date();
+        const expectedLocalDate = cloneTimeStamp.toLocaleDateString("it-IT");
+        const expectedLocalTime = cloneTimeStamp.toLocaleTimeString("it-IT");
+        const conflictEServiceName = `${eService1.name} - clone - ${expectedLocalDate} ${expectedLocalTime}`;
```
Could we revert this change? Same as above.

ecamellini replied:
Same as above

ecamellini commented:
@galales @MalpenZibo @taglioni-r I should have addressed all comments about test flakiness & date inconsistencies, and I tracked the proposed error related improvement in a new task: https://pagopa.atlassian.net/browse/IMN-283
This should be ready for review again :)

taglioni-r commented on packages/catalog-process/test/catalogService.integration.test.ts:
```diff
@@ -1639,139 +1724,248 @@ describe("database test", async () => {
     });
 
     describe("clone descriptor", () => {
-      beforeAll(() => {
-        vi.useFakeTimers();
-        vi.setSystemTime(new Date("2024-01-01T12:00:00"));
-      });
-      afterAll(() => {
-        vi.useRealTimers();
-      });
```
If you add this

```
beforeAll(() => {
    vi.useFakeTimers();
    vi.setSystemTime(new Date());
});
afterAll(() => {
    vi.useRealTimers();
 });
```
you can remove the `{ retry: 3 }` below.
In this way we ensure that every call to `new Date()` both in the logic and in the test will result to the same instant.

ecamellini replied:
True, thanks for the suggestion! Done

galales commented on packages/agreement-process/src/services/agreementService.ts:
```diff
@@ -416,7 +416,13 @@ async function createAndCopyDocumentsForClonedAgreement(
       };
     })
-  );
+  ).catch((error) => {
+    logger.error(
+      `Error copying documents' files for agreement ${clonedAgreement.id} : ${error}`
+    );
+    throw error;
+  });
```
ok from my side to split the activities, as long as the 283 is done right after the current PR, because in my opinion the different error handling is part of this feature.
The sooner we do the refactor, the lesser code will use the old approach and will need to be refactored

galales reviewed (approved):
LGTM, with [this consideration](https://github.com/pagopa/interop-be-monorepo/pull/214#discussion_r1502252934)
